script: Clamp table spans according to the HTML specification

mrobinson opened the pull request:
Previously, spans were partially clamped during layout, but this means
that accessing and setting these properties via script wouldn't behave
according to the HTML specification. In addition, the value wasn't
floored in layout, so could lead to panics. This change improves
clamping and moves it to script.
    
Testing: This change includes a new WPT test.
Fixes #36699.

servo-wpt-sync commented:
🤖 Opened new upstream WPT pull request (web-platform-tests/wpt#52190) with upstreamable changes.

servo-wpt-sync commented:
✍ Updated existing upstream WPT pull request (web-platform-tests/wpt#52190) title and body.

jdm commented:
Wrong issue reference?

Loirooriol commented:
I think this is for #36699

Loirooriol commented on tests/wpt/tests/css/css-tables/colspan-zero-crash.html:
```diff
@@ -0,0 +1,10 @@
+<script>
+  let th = document.createElement("th");
+  span.replaceWith(th);
+  th.colSpan = 0;
```
What if you read `colSpan` after setting? Do browsers say 0 or 1?

servo-wpt-sync commented:
✍ Updated existing upstream WPT pull request (web-platform-tests/wpt#52190) title and body.

simonwuelker reviewed (approved):
~~LGTM~~

simonwuelker commented on components/layout/table/construct.rs:
```diff
@@ -1026,8 +1026,8 @@ where
                     // script).
                     let (rowspan, colspan) = if info.pseudo_element_type.is_none() {
                         let node = info.node.to_threadsafe();
-                        let rowspan = node.get_rowspan().unwrap_or(1).min(65534) as usize;
```
Sorry, I missed this at first: The range is wrong, rowspan can be 0.

https://html.spec.whatwg.org/multipage/tables.html#dom-tdth-rowspan

mrobinson commented:
Yeah, sorry. This needs to be updated as the clamping should happen in script.

mrobinson committed:
script: Clamp table spans according to the HTML specification

Previously, spans were partially clamped during layout, but this means
that accessing and setting these properties via script wouldn't behave
according to the HTML specification. In addition, the value wasn't
floored in layout, so could lead to panics. This change improves
clamping and moves it to script.

Fixes #36699.

Signed-off-by: Martin Robinson <mrobinson@igalia.com>

servo-wpt-sync commented:
✍ Updated existing upstream WPT pull request (web-platform-tests/wpt#52190) title and body.

servo-wpt-sync commented:
✍ Updated existing upstream WPT pull request (web-platform-tests/wpt#52190) title and body.

servo-wpt-sync commented:
📝 Transplanted new upstreamable changes to existing upstream WPT pull request (web-platform-tests/wpt#52190).

mrobinson commented:
Okay. Sorry for the fake out! I think this is ready for review now.

servo-wpt-sync commented:
📝 Transplanted new upstreamable changes to existing upstream WPT pull request (web-platform-tests/wpt#52190).

servo-wpt-sync commented:
📝 Transplanted new upstreamable changes to existing upstream WPT pull request (web-platform-tests/wpt#52190).

Loirooriol commented on tests/wpt/tests/html/semantics/tabular-data/processing-model-1/span-limits.html:
```diff
@@ -63,4 +63,46 @@
+    assert_equals(test5.rowSpan, 65534);
+    assert_equals(test6.rowSpan, 65534);
+}, "rowspan must be clamped to [0, 65534] when parsing attributes");
+</script>
```
Don't duplicate `</script>`

Loirooriol commented on tests/wpt/tests/html/semantics/tabular-data/processing-model-1/span-limits.html:
```diff
@@ -63,4 +63,46 @@
+}, "rowspan must be clamped to [0, 65534] when set via script");
+
+test(() => {
+    document.body.innerHTML += `
```
This will replace all elements with a clone, which seems potentially problematic if e.g. `testharness.js` decides to create the results element before running this test.

Better use `insertAdjacentHTML` or just don't use JS to insert the code (move it to be with the other HTML).

Loirooriol commented on tests/wpt/tests/html/semantics/tabular-data/processing-model-1/col-span-limits.html:
```diff
@@ -56,4 +56,59 @@
+}, "colspan must be clamped to [1, 1000] when set via script");
+
+test(() => {
+    document.body.innerHTML += `
```
Ditto

mrobinson commented:
@Loirooriol Thanks for the review. I think I have addressed your concerns.

servo-wpt-sync commented:
📝 Transplanted new upstreamable changes to existing upstream WPT pull request (web-platform-tests/wpt#52190).

Loirooriol commented on components/script/dom/htmltablecellelement.rs:
```diff
@@ -174,23 +178,26 @@ impl VirtualMethods for HTMLTableCellElement {
+                    // From <https://html.spec.whatwg.org/multipage/#dom-tdth-rowspan>:
+                    // > The rowSpan IDL attribute must reflect the rowspan content attribute. It is clamped to
+                    // > the range [0, 65534], and its default value is 1.
+                    // Note that rowspan = 0 is not supported in quirks mode.
```
But the HTML spec doesn't say to clamp within [1, 65534] in quirks mode. Should it happen during layout instead? Probably needs a test.

mrobinson replied:
Blink and Gecko behave differently here, so this seems under specified:

```html
<body>
    <table>
        <tr>
            <td id="foo" rowspan="0">hello</td>
        </tr>
    </table>
</body>
<script>
  console.log(document.getElementById('foo').rowSpan);
  console.log(document.compatMode);
</script>
```

Loirooriol replied:
Can you file an issue in https://github.com/whatwg/quirks?

mrobinson replied:
I've opened https://github.com/whatwg/quirks/issues/82 for this.

Loirooriol commented on components/script/dom/htmltablecellelement.rs:
```diff
@@ -174,23 +178,26 @@ impl VirtualMethods for HTMLTableCellElement {
+                    // > the range [0, 65534], and its default value is 1.
+                    // Note that rowspan = 0 is not supported in quirks mode.
+                    let document = self.upcast::<Node>().owner_doc();
+                    if document.quirks_mode() != QuirksMode::NoQuirks {
```
Existing problem, but this should be `quirks_mode() == QuirksMode::Quirks`.

Firefox says 0 in limited-quirks mode:

```html
<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Frameset//EN"
   "http://www.w3.org/TR/html4/frameset.dtd">
<table><td id="foo" rowspan="0"></td></table>
<script>document.write(document.getElementById('foo').rowSpan);</script>
```

I see a bunch of `QuirksMode::NoQuirks` in the code that may need to be double-checked, so can be left for a follow-up.

mrobinson commented:
@simonwuelker Do you mind taking another look at this? Thank you!

simonwuelker commented on tests/wpt/meta/html/dom/reflection-tabular.html.ini:
```diff
@@ -7169,9 +7142,6 @@
-    expected: FAIL
-
   [th.colSpan: IDL set to 0]
     expected: FAIL
```
It is suspicious that we don't pass all of these tests after the clamping changes. Have you looked into why that is?

mrobinson replied:
Yes, I took a brief look. It All of these failures seem to be related with how our attribute parsing code is handling values that are outside the limits of certain numerical types. For instance if you grep the metadata for more occurrences of 4294967296 and 2147483648 you will find other tests failing in the same way for other attributes. I suspect we aren't handling this case well for many attributes.

simonwuelker reviewed (approved):
LGTM with one question
